rules/DeadCode/Rector/FunctionLike/NarrowTooWideReturnTypeRector.php

```diff
@@ -0,0 +1,271 @@
+
+        if ($this->shouldSkipNode($node, $scope)) {
+            return null;
+        }
+
+        $returnStatements = $node instanceof ArrowFunction
+            ? []
+            : $this->betterNodeFinder->findReturnsScoped($node);
+        $isAlwaysTerminating = ! $this->silentVoidResolver->hasSilentVoid($node);
+
+        if ($returnStatements === [] && ! $node instanceof ArrowFunction) {
+            return null;
+        }
+
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNode($node);
+        $hasReturnDocblock = (bool) $phpDocInfo?->hasByName('@return');
+
+        if ($hasReturnDocblock) {
+            $returnType = $phpDocInfo->getReturnType();
+        } else {
+            $returnType = $node->returnType;
+            Assert::isInstanceOfAny($returnType, [UnionType::class, NullableType::class]);
+            $returnType = $this->staticTypeMapper->mapPhpParserNodePHPStanType($returnType);
+        }
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+
+        if ($actualReturnTypes === null) {
+            return null;
+        }
+
+        $newReturnType = $this->narrowReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode(
+            $newReturnType,
+            TypeKind::RETURN
+        );
+
+        if ($hasReturnDocblock) {
+            $this->phpDocTypeChanger->changeReturnType($node, $phpDocInfo, $newReturnType);
+        }
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType && ! $returnType instanceof NullableType) {
+            return true;
+        }
+
+        if ($this->hasYield($node)) {
+            return true;
+        }
+
+        if (! $node instanceof ClassMethod) {
+            return false;
+        }
+
+        if ($node->isPrivate() || $node->isFinal()) {
+            return false;
+        }
+
+        if ($node->isAbstract()) {
+            return true;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        if (! $classReflection->isClass()) {
+            return true;
+        }
+
+        return ! $classReflection->isFinalByKeyword();
+    }
+
+    /**
+     * @param Return_[] $returnStatements
+     * @return Type[]|null
+     */
+    private function collectActualReturnTypes(
+        ClassMethod|Function_|Closure|ArrowFunction $node,
+        array $returnStatements,
+        bool $isAlwaysTerminating,
+    ): ?array {
+        if ($node instanceof ArrowFunction) {
+            return [$this->getType($node->expr)];
```

Comment:
use native type check as well, avoid this usage:

```php
class SkipDocblockCallOnArrowFunction
{
    public function run()
    {
        fn (): string|int => $this->execute();
    }

    /**
     * @return string
     */
    private function execute()
    {
        return 1;
    }
}
```

Reply:
Done, how's that?

Also, my apologies on the earlier PR, I'd like to work with you and not against you

Reply:
better to add fixture as well ;)